airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeDataSourceUtils.java

```diff
@@ -0,0 +1,197 @@
+    requestBody.put("refresh_token", refreshCode);
+
+    try {
+      final BodyPublisher bodyPublisher = BodyPublishers.ofString(requestBody.keySet().stream()
+          .map(key -> key + "=" + URLEncoder.encode(requestBody.get(key), StandardCharsets.UTF_8))
+          .collect(joining("&")));
+
+      final byte[] authorization = Base64.getEncoder()
+          .encode((clientId + ":" + clientSecret).getBytes(StandardCharsets.UTF_8));
+
+      final HttpRequest request = HttpRequest.newBuilder()
+          .POST(bodyPublisher)
+          .uri(URI.create(refreshTokenUri))
+          .header("Content-Type", "application/x-www-form-urlencoded")
+          .header("Accept", "application/json")
+          .header("Authorization", "Basic " + new String(authorization, StandardCharsets.UTF_8))
+          .build();
+
+      final HttpResponse<String> response = httpClient.send(request,
+          HttpResponse.BodyHandlers.ofString());
+
+      final JsonNode jsonResponse = Jsons.deserialize(response.body());
+      if (jsonResponse.has("access_token")) {
+        return jsonResponse.get("access_token").asText();
+      } else {
+        throw new RuntimeException(
+            "Failed to obtain accessToken using refresh token. " + jsonResponse);
+      }
+    } catch (final InterruptedException e) {
+      throw new IOException("Failed to refreshToken", e);
+    }
+  }
+
+  public static String buildJDBCUrl(JsonNode config) {
+    final StringBuilder jdbcUrl = new StringBuilder(String.format("jdbc:snowflake://%s/?",
+        config.get("host").asText()));
+
+    // Add required properties
+    jdbcUrl.append(String.format(
+        "role=%s&warehouse=%s&database=%s&schema=%s&JDBC_QUERY_RESULT_FORMAT=%s&CLIENT_SESSION_KEEP_ALIVE=%s",
+        config.get("role").asText(),
+        config.get("warehouse").asText(),
+        config.get("database").asText(),
+        config.get("schema").asText(),
+        // Needed for JDK17 - see
+        // https://stackoverflow.com/questions/67409650/snowflake-jdbc-driver-internal-error-fail-to-retrieve-row-count-for-first-arrow
+        "JSON",
+        true));
+
+    // https://docs.snowflake.com/en/user-guide/jdbc-configure.html#jdbc-driver-connection-string
+    if (config.has("jdbc_url_params")) {
+      jdbcUrl.append("&").append(config.get("jdbc_url_params").asText());
+    }
+    return jdbcUrl.toString();
+  }
+
+  private static Runnable getRefreshTokenTask(final HikariConfig hikariConfig) {
+    return () -> {
+      LOGGER.info("Refresh token process started");
+      var props = hikariConfig.getDataSourceProperties();
+      try {
+        var token = getAccessTokenUsingRefreshToken(props.getProperty("host"),
+            props.getProperty("client_id"), props.getProperty("client_secret"),
+            props.getProperty("refresh_token"));
+        props.setProperty("token", token);
+        hikariConfig.setDataSourceProperties(props);
+        LOGGER.info("New refresh token has been obtained");
```

Comment:
nitpick: is this meant to say `New access token`? (if yes, let's also rename the method to `getAccessTokenTask`)

Reply:
fixed